### DISCUSSION TITLE
Simplifying CellProfiler's input modules

### DISCUSSION THREAD

**Discussion by bethac07** [CellProfiler/CellProfiler#4492] [Ideas]
There is a lot to potentially do here! 

- Upgrading how CellProfiler interacts with its Java VM when it needs to
- Figuring out if there are currently parts of our loading/matching infrastructure that can be simplified for ease of maintenance
- Figuring out where, if anywhere, the current architecture is going to impede our goal of better large image support, and figure out how we can fix that
- Lots of other stuff I'm sure.

**Comment by DavidStirling** [comment]
Here's a summary of my current observations on how things work/what we might want to change:

At present the **Images** module doesn't do too much beyond storing a list of input URLs into a dedicated HDF5 dict class.

The system for **Metadata** is much more complicated but mostly uses Prokaryote/Java. OME-XML metadata keys ('extract from file headers') are handled by extracting the raw XML text and associating that with the file list object. This is then sent into Java during matching. Other user-defined keys are generated by Java-based extractors. These get associated with a Java ImagePlane object. The Python ImagePlaneDetails class wraps the Java ImagePlane objects and these are stored as a list attached to the pipeline.

**NamesAndTypes** is then responsible for image set matching. Before a run (prepare_run) the ImagePlaneDetails list is fetched from the pipeline and Prokaryote comparators are built to perform matching/sorting into image sets. This returns lists of Java ImageSet objects. These get compressed into blobs (possibly with a limited subset of columns/metadata keys?) and stored as a measurement. At the same time a seperate script is run on each ImageSet to extract the metadata keys and store those seperately as Image measurements - note that this is why keys which aren't identical between all images in the set are not available during the rest of the run. The Java ImagePlane object has no dedicated method to fetch custom metadata keys and so there's no obvious way to get keys from an individual image without modifying Prokaryote. 

When NamesAndTypes actually executes during a run the Java ImagePlane object is decompressed from the measurements dict and supplied to the image provider. However, only Series, Index and Channel keys are made available to the provider - these have dedicated fetching methods in the Prokaryote class. Custom keys are effectively unavailable to the provider, which poses a challenge for implementing custom loaders (Bioformats is happy to just use Series & Index but other packages need X, C, T and other info).

**Groups** breaks down the image set list based on metadata stored in the Image measurements dict, so this is largely independent of Prokaryote. Again this suffers from being unable to handle keys which don't match between all images in a set.

**LoadData** is something of an exception in that it appears to use pure Python to manually inject image loaders as requested by the input CSV. This is good news as moving away from prokaryote shouldn't break that functionality.

In terms of how we'd approach restructuring this, I think that most of the matching functionality from prokaryote can be replicated with Python's itertools and other native functionality. We'll have to see how well we can replicate the functionality and how this impacts performance. A pure Python ImagePlaneDetails class which supports arbitrary extra metadata keys should make performing matching much more straightforward.

If we're going to revise the input modules, I reckon this is also a good opportunity to migrate the "Extract from file headers" functionality from Metadata to the Images module. Instead of working with a straight file list I imagine that Images could be responsible for identifying multi-image files and optionally seperating them into specific ImagePlaneDetails objects per series/plane/etc. To ease work with 3D, time series and z-stack images we could provide options to split or not split based on these characteristics (instead of the current all-or-nothing approach which makes working with multiple 3D images within a container quite tricky). Doing such extraction would pre-populate some keys on each ImagePlaneDetails object which can then be passed straight to other modules.

One thing this will probably require will be a Python-based implementation of the OME-XML plane extraction which is currently performed by Prokaryote. Bioformats is also currently used to generate OME-XML files for input images, which isn't a problem but we might want an alternate way to generate and supply required metadata if using other loaders. Metadata would then purely function to extract and append additional metadata keys to Python objects.

**Comment by DavidStirling** [comment]
It's been a little while, so here's a progress update for this project!

We've started work on reconstructing the input modules to move away from java-based implementations. Right now we have the base classes and Images module working, which should set us up for reworking the other modules. Code is currently on the `input_modules` branches of my forks of CellProfiler and Core.

We've put together two new classes which will be used to store and represent images and their metadata:

- `ImageFile` will point to an image's path, url and name. This class also stores a dictionary of image metadata which supports both protected (e.g. SizeC) and user-defined keys. Within the class we have functions to fetch an image reader and extract image metadata from file headers. This is currently hard-coded to use BioFormats, but in time the plan is to abstract this out and support additional image loaders. The Metadata module will attach keys to the ImageFile's metadata dictionary.

- `ImagePlane` represents an individual 2D plane within an image (eg. Z=2, C=1, T=0). each `ImagePlane` object is linked to it's parent `ImageFile`. It's metadata store inherits from the parent ImageFile but also supports plane-level keys. This will allow for plane indexing and metadata retrieval without needing to call javabridge. This class will be used to generate groups in NamesAndTypes.

The new Images module holds the same basic functionality as the original module, but now performs metadata extraction from file headers if desired (previously an option in the Metadata module). In practice this will also resolve the outstanding issues with CellProfiler getting confused over whether metadata was already extracted, since this is now stored as readable info on the python class instead of in a java object. Images is also now responsible for generating `ImagePlane` objects from each `ImageFile`, as determined by the new settings to split individual dimensions. This will provide more granularity in working with n-dimensional images, where previously you could either split everything or nothing. We may want to consider having the 3D mode toggle in this module too. My hope is that this will make things more extensible in the future if we want to support time-series analysis or tiling of large images.

To support all of this I've revised the file list display to be a 'real' TreeCtrl object, with support for displaying the series info using metadata extracted from each image (if enabled). You can see that in the screenshot below:


<img width="1193" alt="Screenshot 2022-02-22 at 09 43 46" src="https://user-images.githubusercontent.com/26802537/155106177-d9ac271b-7fd1-4ddc-8379-06930214c442.png">

In terms of structural changes, the `Workspace` class remains largely the same, but the `Pipeline` class is changing. In the current release `workspace.file_list` stores a list of URLs to images and `pipeline.file_list` copies the workspace list (most of the time). In the new implementation `pipeline.file_list` will instead be a list of `ImageFile` objects. Filtering of the file list (in `prepare_run`) is now performed in Python rather than over Java. `pipeline.image_plane_list` holds the list of image plane objects.

Hopefully this is useful, feedback is more than welcome!

**Comment by DavidStirling** [comment]
March/April progress update:

Things have come along a fair bit since my last post. We now have all 4 input modules running without Java using the ImageFile-ImagePlane hierarchy mentioned previously. Metadata and NamesAndTypes shouldn't look all that different since most changes are under the hood, though I've made some small improvements to the image set/metadata table display to make it easier when working with multi-plane files.

There's a lot of testing still to be done, plus a need for a new test suite. I expect that we largely have feature parity with the java-based modules, key areas to check will be obscure scenarios e.g. loading Objects from an image URL.

As a part of this work we've revised the way file lists are stored in the HDF5 measurements dict/.cpproj file. Previously the list was generated with a group representing each subfolder. All image names in a folder were stored as a single concatenated string byte array with a seperate indexing array to mark the start/end of each file, plus an additional metadata table to hold OME-XML data as strings. The latter table wasn't really utilised for anything more than moving between machines without having to re-extract file headers.

The new HDF5 Filelist system makes a few key revisions:
- Image paths are no longer escaped/encoded when generating groups. This was leftover functionality from Python 2 which is no longer necessary, the native interface now seems fine with Unicode strings.
- Filenames within a folder are stored as a (sorted) variable length string array instead of being concatenated into a long byte array. This is easier to read and manipulate, plus we can lose the indexing array. The older implementation used a custom VStringArray class which dated back to before hdf5 supported variable length content, so that was overcomplicated for what we need to achieve here.
- Instead of storing the entire OME-XML header, we now store metadata as a numeric array. Each series within a file with extracted metadata is stored as 5 integers representing T,C,Z,Y and X dimensions, with each series stored sequentially so that we have 1 array per image. This is substantially more space efficient while being equally informative, and more importantly allows non-bioformats readers to supply dimensional metadata even if OME-XML data is absent.
- Slightly related, I've fixed the analysis startup sequence to minimise individual writes to the measurements dict and avoid copying the entire file list when transmitting the pipeline. This prevents that issue where CP4 hangs for a minute when starting analysis with a large file list.

The changes outlined here naturally break compatibility with older .cpproj files. I've set it up to ignore old-format project file lists and just load the pipeline from those files. Due to this we decided to just increment version number to CP 5.0.0 (beta), which will more readily designate compatible vs incompatible pipelines during development.

The other major change has been to image readers. The current system has been to have a `get_image_reader` function that calls python-bioformats after checking against a list of possible exceptions for inputs like OMERO ids. This has been getting increasingly messy as we add more image types.

To alleviate this, we've put together a modular image reader system based on the new [cellprofiler_core.reader](https://github.com/DavidStirling/core/blob/input_modules/cellprofiler_core/reader/_reader.py) abstract base class. I've migrated bioformats functionality into one of these reader classes (pretty straightforward as the reader class largely mirrors the bioformats API). We also now have a basic imageio reader for some simple file types, since imageio was already a dependency and we needed another reader for testing purposes. In a seperate project we'll eventually move towards adding a zarr/NGFF reader under this system.

The current system for selecting a reader involves the `supports_format` class method. When provided with a file name this is expected to return a "competency value" between 1 and 4 reflecting whether the reader is optimal (1) or terrible (4) at reading a format (or anything else if it's entirely incompatible). A value of 1 will trigger CellProfiler to stop searching and use that reader, otherwise the best scoring reader is selected once all are tested. For the most part this test should be done just by looking at the file extension. For example, if we want to open a .png file we query both our BioFormats and ImageIO readers. BioFormats typically returns '3', since while it's powerful it needs Java. ImageIO returns '2' since it's very good with PNG files, this means that ImageIO is selected and used from here on. The selected reader type is cached with the image metadata on the ImageFile class, so a file should only be tested once. I'll be adding a preferences option to force CP to always use BioFormats, so that the old system can be used if people want it.

Eventually I'd like to manage readers in a similar way to Module plugins, so users can add their own readers to their plugins folder and they'll load appropriately.

The end goal from Glencoe's end has been to avoid needing to start java vms when running CellProfiler without bioformats as the primary reader (i.e. with NGFF). To that end java is now set up to be initialised only when it's first called. Javabridge code should now only be called from within the Python-Bioformats package, which will hopefully make things easier if there are plans to migrate that to a scyjava implementation.

@bethac07 One outstanding issue is that SaveImages currently uses BioFormats for writing images, and it's the only remaining module with a hard java dependency. Given that the supported formats in SaveImages are limited to simple types, I'd like us to consider replacing the BioFormats writer with skimage.io if we can achieve the same output results?

**Comment by DavidStirling** [comment]
April/May progress update:

I think we're nearly there! Most work this month has been on bug fixing and testing.

The input modules appear to be working as expected and I believe the obscure use cases are now handled correctly. The test suite has been revised to account for the changes made on the branch.

Readers can now be provided as plugins should the user wish! The plugins folder is now scanned for both Module and Reader classes to load.

The big roadblock has been getting java to launch on demand. While this worked nicely in Test mode, Analysis mode presented some challenges. After a lot of digging it turned out that javabridge struggles to start it's monitoring thread if another thread is currently holding an active connection to STDIN. As it happens, Workers were set up with a system which connects to STDIN on a thread and uses STDIN being closed as a signal that the parent process has died (and thus to shut down). This unusual approach to "has CellProfiler crashed?" turned out to be incompatible with on-demand Java and so we looked for another (more pythonic?) way to perform this function.

This led to needing to figure out exactly how the ZMQ inter-process communication system works. Surprisingly much of that system is actually nonfunctional in CP4, Analysis mode working at all was very lucky. We had a bunch of unused sockets lying around and I'm pretty sure that the Knime bridge does nothing at all. Analysis mode was also clearly designed with the intention of launching multiple analysis runs at once, which modern CP absolutely can't do. Because of this there was a lot of back and forth communication which serves no real purpose. Thus as part of a refactor to remove the reliance on STDIN connections we've tried to clean this up a bit.

It's a lot to get across over text, but I've put together a [Wiki page](https://github.com/CellProfiler/CellProfiler/wiki/Analysis-Mode-and-ZMQ-Interfaces) documenting the new system. The key changes are:
- STDIN is no longer used as a run/shutdown signal.
- Instead, the boundary thread periodically broadcasts a keepalive signal (heartbeat) to tell workers that the main process is alive.
- That same socket is also used to deliver an explicit shutdown signal, after which the worker has 15s to exit gracefully. In CP4 STDIN closure is used to trigger shutdown under normal conditions, with a 10s timeout before hard-kill. In practice it didn't really work, which is why there's often a long delay before an analysis run stops.
- Analysis ID and work address dictionaries are no longer provided via a repeating broadcast from the boundary. Previously each analysis would add another entry to the dictionary broadcast which was never removed, causing slowdown over time. The idea was to have the worker look up the analysis which it's been assigned, but with only one analysis running it was unnecessary.
- Instead, the worker is given a single analysis ID and connection address when launched. This eliminates the need to continually re-check analysis status between jobs, which should reduce delay between imagesets. 
- ZMQ context/sockets are refreshed for each analysis run, which should address the random freezes sometimes seen in Analysis mode in CP4.

Future work to be considered (for different PRs since the current one is already huge):
- Add an NGFF .zarr reader
- Split OMERO server reading off from BioFormats into it's own dedicated reader, possibly using OMERO-py.
- Replacing the worker logging system. Right now a dedicated thread is created for each worker to read it's STDOUT and log it to console. ZMQ has a PUBHandler class which could easily be used to broadcast logging straight back to a single socket on one thread. This would cut down load on the main process and make timestamps more accurate.
- Add a preference for "allow opening of files" when selecting readers, if we find that readers actually need to do this.
- Remove the rest of Prokaryote/eventually remove prokaryote from builds.
- Add a full testing suite for image readers.
